org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1582 @@
+    private BufferProxy write_buf;
+    private SSLFDProxy ssl_fd;
+
+    private PK11Cert cert;
+    private PK11PrivKey key;
+    private X509KeyManager[] key_managers;
+    private X509TrustManager[] trust_managers;
+
+    private boolean need_client_auth;
+    private boolean want_client_auth;
+
+    private SSLEngineResult.HandshakeStatus handshake_state = SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING;
+
+    private SSLCipher[] enabled_ciphers;
+    private SSLVersion min_protocol;
+    private SSLVersion max_protocol;
+
+    private JSSSession session;
+    private int unknown_state_count;
+
+    private boolean step_handshake;
+
+    private boolean is_outbound_closed = false;
+    private boolean is_inbound_closed = false;
+
+    private SSLException ssl_exception;
+    private boolean seen_exception;
+
+    private int debug_port;
+    private ServerSocket ss_socket;
+    private Socket s_socket;
+    private Socket c_socket;
+    private OutputStream s_ostream;
+    private OutputStream c_ostream;
+
+    private String name;
+    private String prefix = "";
+
+    public JSSEngine() {
+        peer_info = "";
+        session = new JSSSession(BUFFER_SIZE);
+        debug("JSSEngine: constructor()");
+    }
+
+    public JSSEngine(String peerHost, int peerPort) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerPort;
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+        debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ")");
+    }
+
+    public JSSEngine(String peerHost, int peerPort,
+                     org.mozilla.jss.crypto.X509Certificate localCert,
+                     org.mozilla.jss.crypto.PrivateKey localKey) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerHost;
+        cert = (PK11Cert) localCert;
+        key = (PK11PrivKey) localKey;
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+
+        prefix = prefix + "[" + peer_info + "] ";
+
+        debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ", " + localCert + ", " + localKey + ")");
+    }
+
+    private String errorText(int error) {
+        // Convert the given error into a pretty string representation with
+        // as much information as is currently available.
+
+        String error_name = PR.ErrorToName(error);
+        String error_text = PR.GetErrorText();
+
+        if (error == 0) {
+            return "NO ERROR";
+        } else if (error_name.isEmpty()) {
+            return "UNKNOWN (" + error + ")";
+        } else if (error_text.isEmpty()) {
+            return error_name + " (" + error + ")";
+        } else {
+            return error_name + " (" + error + "): " + error_text;
+        }
+    }
+
+    private void debug(String msg) {
+        logger.debug(prefix + msg);
```

Comment:
It depends on how you use the logging. If the logging is configured to show the source line numbers, this encapsulation will cause it to show the same line number for all log messages. If the line numbers are not shown it doesn't really matter.

Reply:
Ah, I wasn't aware that it was possible to show line numbers. I'll collapse this down into regular annotations.

Having the prefixed information really helps debugging when you're using more than one SSLEngine. Do you know if it is possible to add additional prefixes to the logging context in a class?

Reply:
You probably can create a separate logger instance for each JSSEngine instance instead of using a static one:
```
this.logger = LoggerFactory.getLogger(JSSEngine.class.getName() + "/" + engineName);
```
There is only limited info that we can record. See:
* https://docs.oracle.com/javase/8/docs/api/java/util/logging/LogRecord.html
* https://github.com/dogtagpki/pki/blob/master/base/util/src/org/dogtagpki/util/logging/PKIFormatter.java